// @electron-forge/plugin-webpack/dist/WebpackPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _asyncOra = require("@electron-forge/async-ora");

var _pluginBase = _interopRequireDefault(require("@electron-forge/plugin-base"));

var _webMultiLogger = _interopRequireDefault(require("@electron-forge/web-multi-logger"));

var _debug = _interopRequireDefault(require("debug"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _path = _interopRequireDefault(require("path"));

var _webpack = _interopRequireDefault(require("webpack"));

var _webpackDevServer = _interopRequireDefault(require("webpack-dev-server"));

var _ElectronForgeLogging = _interopRequireDefault(require("./util/ElectronForgeLogging"));

var _once = _interopRequireDefault(require("./util/once"));

var _WebpackConfig = _interopRequireDefault(require("./WebpackConfig"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const d = (0, _debug.default)('electron-forge:plugin:webpack');
const DEFAULT_PORT = 3000;
const DEFAULT_LOGGER_PORT = 9000;

class WebpackPlugin extends _pluginBase.default {
  constructor(c) {
    super(c);

    _defineProperty(this, "name", 'webpack');

    _defineProperty(this, "isProd", false);

    _defineProperty(this, "projectDir", void 0);

    _defineProperty(this, "baseDir", void 0);

    _defineProperty(this, "_configGenerator", void 0);

    _defineProperty(this, "watchers", []);

    _defineProperty(this, "servers", []);

    _defineProperty(this, "loggers", []);

    _defineProperty(this, "port", DEFAULT_PORT);

    _defineProperty(this, "loggerPort", DEFAULT_LOGGER_PORT);

    _defineProperty(this, "isValidPort", port => {
      if (port < 1024) {
        throw new Error(`Cannot specify port (${port}) below 1024, as they are privileged`);
      } else if (port > 65535) {
        throw new Error(`Port specified (${port}) is not a valid TCP port.`);
      } else {
        return true;
      }
    });

    _defineProperty(this, "exitHandler", (options, err) => {
      d('handling process exit with:', options);

      if (options.cleanup) {
        for (const watcher of this.watchers) {
          d('cleaning webpack watcher');
          watcher.close(() => {});
        }

        this.watchers = [];

        for (const server of this.servers) {
          d('cleaning http server');
          server.close();
        }

        this.servers = [];

        for (const logger of this.loggers) {
          d('stopping logger');
          logger.stop();
        }

        this.loggers = [];
      }

      if (err) console.error(err.stack);
      if (options.exit) process.exit();
    });

    _defineProperty(this, "runWebpack", async (options, isRenderer = false) => new Promise((resolve, reject) => {
      (0, _webpack.default)(options).run(async (err, stats) => {
        if (isRenderer && this.config.renderer.jsonStats) {
          await this.writeJSONStats('renderer', stats, options.stats);
        }

        if (err) {
          return reject(err);
        }

        return resolve(stats);
      });
    }));

    _defineProperty(this, "init", dir => {
      this.setDirectories(dir);
      d('hooking process events');
      process.on('exit', _code => this.exitHandler({
        cleanup: true
      }));
      process.on('SIGINT', _signal => this.exitHandler({
        exit: true
      }));
    });

    _defineProperty(this, "setDirectories", dir => {
      this.projectDir = dir;
      this.baseDir = _path.default.resolve(dir, '.webpack');
    });

    _defineProperty(this, "loggedOutputUrl", false);

    _defineProperty(this, "resolveForgeConfig", async forgeConfig => {
      if (!forgeConfig.packagerConfig) {
        forgeConfig.packagerConfig = {};
      }

      if (forgeConfig.packagerConfig.ignore) {
        console.error(`You have set packagerConfig.ignore, the Electron Forge webpack plugin normally sets this automatically.

Your packaged app may be larger than expected if you dont ignore everything other than the '.webpack' folder`.red);
        return forgeConfig;
      }

      forgeConfig.packagerConfig.ignore = file => {
        if (!file) return false;

        if (this.config.jsonStats && file.endsWith(_path.default.join('.webpack', 'main', 'stats.json'))) {
          return true;
        }

        if (this.config.renderer.jsonStats && file.endsWith(_path.default.join('.webpack', 'renderer', 'stats.json'))) {
          return true;
        }

        return !/^[/\\]\.webpack($|[/\\]).*$/.test(file);
      };

      return forgeConfig;
    });

    _defineProperty(this, "packageAfterCopy", async (_, buildPath) => {
      const pj = await _fsExtra.default.readJson(_path.default.resolve(this.projectDir, 'package.json'));

      if (pj.config) {
        delete pj.config.forge;
      }

      pj.devDependencies = {};
      pj.dependencies = {};
      pj.optionalDependencies = {};
      pj.peerDependencies = {};
      await _fsExtra.default.writeJson(_path.default.resolve(buildPath, 'package.json'), pj, {
        spaces: 2
      });
      await _fsExtra.default.mkdirp(_path.default.resolve(buildPath, 'node_modules'));
    });

    _defineProperty(this, "compileMain", async (watch = false, logger) => {
      let tab;

      if (logger) {
        tab = logger.createTab('Main Process');
      }

      await (0, _asyncOra.asyncOra)('Compiling Main Process Code', async () => {
        const mainConfig = await this.configGenerator.getMainConfig();
        await new Promise((resolve, reject) => {
          const compiler = (0, _webpack.default)(mainConfig);
          const [onceResolve, onceReject] = (0, _once.default)(resolve, reject);

          const cb = async (err, stats) => {
            if (tab && stats) {
              tab.log(stats.toString({
                colors: true
              }));
            }

            if (this.config.jsonStats) {
              await this.writeJSONStats('main', stats, mainConfig.stats);
            }

            if (err) return onceReject(err);

            if (!watch && stats !== null && stats !== void 0 && stats.hasErrors()) {
              return onceReject(new Error(`Compilation errors in the main process: ${stats.toString()}`));
            }

            return onceResolve(undefined);
          };

          if (watch) {
            this.watchers.push(compiler.watch({}, cb));
          } else {
            compiler.run(cb);
          }
        });
      });
    });

    _defineProperty(this, "compileRenderers", async (watch = false) => {
      await (0, _asyncOra.asyncOra)('Compiling Renderer Template', async () => {
        const stats = await this.runWebpack(await this.configGenerator.getRendererConfig(this.config.renderer.entryPoints), true);

        if (!watch && stats !== null && stats !== void 0 && stats.hasErrors()) {
          throw new Error(`Compilation errors in the renderer: ${stats.toString()}`);
        }
      });

      for (const entryPoint of this.config.renderer.entryPoints) {
        if (entryPoint.preload) {
          await (0, _asyncOra.asyncOra)(`Compiling Renderer Preload: ${entryPoint.name}`, async () => {
            await this.runWebpack(await this.configGenerator.getPreloadRendererConfig(entryPoint, entryPoint.preload));
          });
        }
      }
    });

    _defineProperty(this, "launchDevServers", async logger => {
      await (0, _asyncOra.asyncOra)('Launch Dev Servers', async () => {
        const tab = logger.createTab('Renderers');
        const pluginLogs = new _ElectronForgeLogging.default(tab);
        const config = await this.configGenerator.getRendererConfig(this.config.renderer.entryPoints);
        if (!config.plugins) config.plugins = [];
        config.plugins.push(pluginLogs);
        const compiler = (0, _webpack.default)(config);
        const webpackDevServer = new _webpackDevServer.default(compiler, {
          hot: true,
          port: this.port,
          static: _path.default.resolve(this.baseDir, 'renderer'),
          devMiddleware: {
            writeToDisk: true
          },
          setupExitSignals: true,
          historyApiFallback: true
        });
        const server = await webpackDevServer.listen(this.port);
        this.servers.push(server);
      });
      await (0, _asyncOra.asyncOra)('Compiling Preload Scripts', async () => {
        for (const entryPoint of this.config.renderer.entryPoints) {
          if (entryPoint.preload) {
            const config = await this.configGenerator.getPreloadRendererConfig(entryPoint, entryPoint.preload);
            await new Promise((resolve, reject) => {
              const tab = logger.createTab(`${entryPoint.name} - Preload`);
              const [onceResolve, onceReject] = (0, _once.default)(resolve, reject);
              this.watchers.push((0, _webpack.default)(config).watch({}, (err, stats) => {
                if (stats) {
                  tab.log(stats.toString({
                    colors: true
                  }));
                }

                if (err) return onceReject(err);
                return onceResolve(undefined);
              }));
            });
          }
        }
      });
    });

    _defineProperty(this, "alreadyStarted", false);

    if (c.port) {
      if (this.isValidPort(c.port)) {
        this.port = c.port;
      }
    }

    if (c.loggerPort) {
      if (this.isValidPort(c.loggerPort)) {
        this.loggerPort = c.loggerPort;
      }
    }

    this.startLogic = this.startLogic.bind(this);
    this.getHook = this.getHook.bind(this);
  }

  async writeJSONStats(type, stats, statsOptions) {
    if (!stats) return;
    d(`Writing JSON stats for ${type} config`);
    const jsonStats = stats.toJson(statsOptions);

    const jsonStatsFilename = _path.default.resolve(this.baseDir, type, 'stats.json');

    await _fsExtra.default.writeJson(jsonStatsFilename, jsonStats, {
      spaces: 2
    });
  } // eslint-disable-next-line max-len


  get configGenerator() {
    // eslint-disable-next-line no-underscore-dangle
    if (!this._configGenerator) {
      // eslint-disable-next-line no-underscore-dangle
      this._configGenerator = new _WebpackConfig.default(this.config, this.projectDir, this.isProd, this.port);
    } // eslint-disable-next-line no-underscore-dangle


    return this._configGenerator;
  }

  getHook(name) {
    switch (name) {
      case 'prePackage':
        this.isProd = true;
        return async () => {
          await _fsExtra.default.remove(this.baseDir);
          await this.compileMain();
          await this.compileRenderers();
        };

      case 'postStart':
        return async (_, child) => {
          if (!this.loggedOutputUrl) {
            console.info(`\n\nWebpack Output Available: ${`http://localhost:${this.loggerPort}`.cyan}\n`);
            this.loggedOutputUrl = true;
          }

          d('hooking electron process exit');
          child.on('exit', () => {
            if (child.restarted) return;
            this.exitHandler({
              cleanup: true,
              exit: true
            });
          });
        };

      case 'resolveForgeConfig':
        return this.resolveForgeConfig;

      case 'packageAfterCopy':
        return this.packageAfterCopy;

      default:
        return null;
    }
  }

  async startLogic() {
    if (this.alreadyStarted) return false;
    this.alreadyStarted = true;
    await _fsExtra.default.remove(this.baseDir);
    const logger = new _webMultiLogger.default(this.loggerPort);
    this.loggers.push(logger);
    await this.compileMain(true, logger);
    await this.launchDevServers(logger);
    await logger.start();
    return false;
  }

}

exports.default = WebpackPlugin;
