// @electron-forge/plugin-webpack/dist/WebpackConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _debug = _interopRequireDefault(require("debug"));

var _htmlWebpackPlugin = _interopRequireDefault(require("html-webpack-plugin"));

var _path = _interopRequireDefault(require("path"));

var _webpack = _interopRequireDefault(require("webpack"));

var _webpackMerge = require("webpack-merge");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const d = (0, _debug.default)('electron-forge:plugin:webpack:webpackconfig');

class WebpackConfigGenerator {
  constructor(pluginConfig, projectDir, isProd, port) {
    _defineProperty(this, "isProd", void 0);

    _defineProperty(this, "pluginConfig", void 0);

    _defineProperty(this, "port", void 0);

    _defineProperty(this, "projectDir", void 0);

    _defineProperty(this, "webpackDir", void 0);

    this.pluginConfig = pluginConfig;
    this.projectDir = projectDir;
    this.webpackDir = _path.default.resolve(projectDir, '.webpack');
    this.isProd = isProd;
    this.port = port;
    d('Config mode:', this.mode);
  }

  resolveConfig(config) {
    if (typeof config === 'string') {
      // eslint-disable-next-line import/no-dynamic-require, global-require
      return require(_path.default.resolve(this.projectDir, config));
    }

    return config;
  }

  get mode() {
    return this.isProd ? 'production' : 'development';
  }

  rendererEntryPoint(entryPoint, inRendererDir, basename) {
    if (this.isProd) {
      return `\`file://$\{require('path').resolve(__dirname, '..', '${inRendererDir ? 'renderer' : '.'}', '${entryPoint.name}', '${basename}')}\``;
    }

    const baseUrl = `http://localhost:${this.port}/${entryPoint.name}`;

    if (basename !== 'index.html') {
      return `'${baseUrl}/${basename}'`;
    }

    return `'${baseUrl}'`;
  }

  toEnvironmentVariable(entryPoint, preload = false) {
    const suffix = preload ? '_PRELOAD_WEBPACK_ENTRY' : '_WEBPACK_ENTRY';
    return `${entryPoint.name.toUpperCase().replace(/ /g, '_')}${suffix}`;
  }

  getPreloadDefine(entryPoint) {
    if (entryPoint.preload) {
      if (this.isProd) {
        return `require('path').resolve(__dirname, '../renderer', '${entryPoint.name}', 'preload.js')`;
      }

      return `'${_path.default.resolve(this.webpackDir, 'renderer', entryPoint.name, 'preload.js').replace(/\\/g, '\\\\')}'`;
    } // If this entry-point has no configured preload script just map this constant to `undefined`
    // so that any code using it still works.  This makes quick-start / docs simpler.


    return 'undefined';
  }

  assetRelocatorBaseDir(inRendererDir = true) {
    if (this.isProd) {
      return `process.resourcesPath + "/" + (__filename.includes(".asar") ? "app.asar" : "app") + "/.webpack/${inRendererDir ? 'main' : 'renderer/any_folder'}"`;
    }

    return JSON.stringify(_path.default.resolve(this.webpackDir, inRendererDir ? 'main' : 'renderer', inRendererDir ? '.' : 'any_folder'));
  }

  getDefines(inRendererDir = true) {
    const defines = {
      ASSET_RELOCATOR_BASE_DIR: this.assetRelocatorBaseDir(inRendererDir)
    };

    if (!this.pluginConfig.renderer.entryPoints || !Array.isArray(this.pluginConfig.renderer.entryPoints)) {
      throw new Error('Required config option "renderer.entryPoints" has not been defined');
    }

    for (const entryPoint of this.pluginConfig.renderer.entryPoints) {
      const entryKey = this.toEnvironmentVariable(entryPoint);

      if (entryPoint.html) {
        defines[entryKey] = this.rendererEntryPoint(entryPoint, inRendererDir, 'index.html');
      } else {
        defines[entryKey] = this.rendererEntryPoint(entryPoint, inRendererDir, 'index.js');
      }

      defines[`process.env.${entryKey}`] = defines[entryKey];
      const preloadDefineKey = this.toEnvironmentVariable(entryPoint, true);
      defines[preloadDefineKey] = this.getPreloadDefine(entryPoint);
      defines[`process.env.${preloadDefineKey}`] = defines[preloadDefineKey];
    }

    return defines;
  }

  sourceMapOption() {
    return this.isProd ? 'source-map' : 'eval-source-map';
  }

  getMainConfig() {
    const mainConfig = this.resolveConfig(this.pluginConfig.mainConfig);

    if (!mainConfig.entry) {
      throw new Error('Required option "mainConfig.entry" has not been defined');
    }

    const fix = item => {
      if (typeof item === 'string') return fix([item])[0];

      if (Array.isArray(item)) {
        return item.map(val => val.startsWith('./') ? _path.default.resolve(this.projectDir, val) : val);
      }

      const ret = {};

      for (const key of Object.keys(item)) {
        ret[key] = fix(item[key]);
      }

      return ret;
    };

    mainConfig.entry = fix(mainConfig.entry);
    return (0, _webpackMerge.merge)({
      devtool: 'source-map',
      target: 'electron-main',
      mode: this.mode,
      output: {
        path: _path.default.resolve(this.webpackDir, 'main'),
        filename: 'index.js',
        libraryTarget: 'commonjs2'
      },
      plugins: [new _webpack.default.DefinePlugin(this.getDefines())],
      node: {
        __dirname: false,
        __filename: false
      },
      resolve: {
        modules: [_path.default.resolve(_path.default.dirname(this.webpackDir), './'), _path.default.resolve(_path.default.dirname(this.webpackDir), 'node_modules'), _path.default.resolve(__dirname, '..', 'node_modules')]
      }
    }, mainConfig || {});
  }

  async getPreloadRendererConfig(parentPoint, entryPoint) {
    const rendererConfig = this.resolveConfig(this.pluginConfig.renderer.config);
    const prefixedEntries = entryPoint.prefixedEntries || [];
    return (0, _webpackMerge.merge)({
      devtool: this.sourceMapOption(),
      mode: this.mode,
      entry: prefixedEntries.concat([entryPoint.js]),
      output: {
        path: _path.default.resolve(this.webpackDir, 'renderer', parentPoint.name),
        filename: 'preload.js'
      },
      node: {
        __dirname: false,
        __filename: false
      }
    }, rendererConfig || {}, {
      target: 'electron-preload'
    });
  }

  async getRendererConfig(entryPoints) {
    const rendererConfig = this.resolveConfig(this.pluginConfig.renderer.config);
    const entry = {};

    for (const entryPoint of entryPoints) {
      const prefixedEntries = entryPoint.prefixedEntries || [];
      entry[entryPoint.name] = prefixedEntries.concat([entryPoint.js]);
    }

    const defines = this.getDefines(false);
    const plugins = entryPoints.filter(entryPoint => Boolean(entryPoint.html)).map(entryPoint => new _htmlWebpackPlugin.default({
      title: entryPoint.name,
      template: entryPoint.html,
      filename: `${entryPoint.name}/index.html`,
      chunks: [entryPoint.name].concat(entryPoint.additionalChunks || [])
    })).concat([new _webpack.default.DefinePlugin(defines)]);
    return (0, _webpackMerge.merge)({
      entry,
      devtool: this.sourceMapOption(),
      target: ['web', 'electron-renderer'],
      mode: this.mode,
      output: {
        path: _path.default.resolve(this.webpackDir, 'renderer'),
        filename: '[name]/index.js',
        globalObject: 'self',
        ...(this.isProd ? {} : {
          publicPath: '/'
        })
      },
      node: {
        __dirname: false,
        __filename: false
      },
      plugins
    }, rendererConfig || {});
  }

}

exports.default = WebpackConfigGenerator;
