// @electron-forge/plugin-webpack/dist/util/ElectronForgeLogging.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _asyncOra = require("@electron-forge/async-ora");

var _once = _interopRequireDefault(require("./once"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const pluginName = 'ElectronForgeLogging';

class LoggingPlugin {
  constructor(tab) {
    _defineProperty(this, "tab", void 0);

    _defineProperty(this, "promiseResolver", void 0);

    _defineProperty(this, "promiseRejector", void 0);

    this.tab = tab;
    this.promiseResolver = undefined;
    this.promiseRejector = undefined;
  }

  addRun() {
    if (this.promiseResolver) this.promiseResolver();
    (0, _asyncOra.asyncOra)('Compiling Renderer Code', () => new Promise((resolve, reject) => {
      const [onceResolve, onceReject] = (0, _once.default)(resolve, reject);
      this.promiseResolver = onceResolve;
      this.promiseRejector = onceReject;
    }), () => {});
  }

  finishRun(error) {
    if (error && this.promiseRejector) this.promiseRejector(error);else if (this.promiseResolver) this.promiseResolver();
    this.promiseRejector = undefined;
    this.promiseResolver = undefined;
  }

  apply(compiler) {
    compiler.hooks.watchRun.tap(pluginName, _compiler => {
      this.addRun();
    });
    compiler.hooks.done.tap(pluginName, stats => {
      if (stats) {
        this.tab.log(stats.toString({
          colors: true
        }));

        if (stats.hasErrors()) {
          this.finishRun(stats.compilation.getErrors().toString());
          return;
        }
      }

      this.finishRun();
    });
    compiler.hooks.failed.tap(pluginName, err => this.finishRun(err.message));
    compiler.hooks.infrastructureLog.tap(pluginName, (name, _type, args) => {
      this.tab.log(`${name} - ${args.join(' ')}\n`);
      return true;
    });
  }

}

exports.default = LoggingPlugin;
